src/3d/qgscameracontroller.cpp

```diff
@@ -556,31 +565,26 @@ void QgsCameraController::handleTerrainNavigationWheelZoom()
 
   // First transformation : Shift camera position and view center and rotate the camera
   {
-    QVector3D shiftVector = mZoomPoint - mCamera->viewCenter();
-
-    QVector3D newViewCenterWorld = camera()->viewCenter() + shiftVector;
+    QVector3D shiftVector = mZoomPoint - mCameraBeforeZoom->position();
+    QVector3D newViewCenterWorld = mCameraBeforeZoom->position() + shiftVector;
 
     mCameraPose.setDistanceFromCenterPoint( dist );
     mCameraPose.setCenterPoint( newViewCenterWorld );
-    updateCameraFromPose();
+    mCameraPose.updateCamera( mCamera );
   }
 
   // Second transformation : Shift camera position back
   {
     QgsRay3D ray = Qgs3DUtils::rayFromScreenPoint( QPoint( mMousePos.x(), mMousePos.y() ), mViewport.size(), mCamera );
     QVector3D clickedPositionWorld = ray.origin() + dist * ray.direction();
 
-    QVector3D shiftVector = clickedPositionWorld - mCamera->viewCenter();
-
+    QVector3D shiftVector = clickedPositionWorld - camera()->viewCenter();
     QVector3D newViewCenterWorld = camera()->viewCenter() - shiftVector;
-    QVector3D newCameraPosition = camera()->position() - shiftVector;
 
-    mCameraPose.setDistanceFromCenterPoint( ( newViewCenterWorld - newCameraPosition ).length() );
+    mCameraPose.setDistanceFromCenterPoint( dist );
     mCameraPose.setCenterPoint( newViewCenterWorld );
     updateCameraFromPose();
```

Comment:
There is very similar code for zoom in/out in onPositionChangedTerrainNavigation() - should that code be adjusted as well?